Improvements to Out-of-Core Hash Join

lnkuiper opened the pull request:
Continued where I left off. We now partition the probe side as well (instead of scanning the whole probe side multiple times), which means we can support out-of-core right/outer/mark/anti joins.

Probe-side partitioning happens during the streaming probe (`PhysicalHashJoin::Execute`): tuples that cannot be probed against the partitions of the hash table during this phase, because they belong to different partitions, are sunk into a `PartitionedColumnData`. However, if we can finish probing in just two rounds, the probe-side does not need to be partitioned at all, and we sink into a `ColumnDataCollection` instead.

`PartitionedColumnData` is a new generic partitioning interface. I've implemented `RadixPartitionedColumnData`. An issue with multi-threaded partitioning in general is that each thread allocates data for all partitions, which results in high memory usage. We prevent this by sharing a single allocator per partition across all threads. The `PartitionedColumnData` class can be extended to do, e.g., Hive partitioning in a streaming manner. 

I've also implemented the `ColumnDataConsumer` class, which can read and consume a `ColumnDataCollection`. This is useful for the out-of-core hash join, as we wish to read the probe side `ColumnDataCollection` just once. Previously, the read data was written back to disk before being thrown away, which was wasteful.

I've run the same benchmark as in my previous PR on my laptop, which is joining two tables with 100M integers each, but they only have 1k matches. Here are the numbers:

| Memory limit (GB) | Old time (s) | New time (s) |
|:-|:-|:-|
|10|1.97|1.96|
|9|1.97|1.97|
|8|2.23|2.22|
|7|2.23|2.44|
|6|2.27|2.39|
|5|2.27|2.32|
|4|2.81|2.45|
|3|5.60|3.20|
|2|7.69|3.28|
|1|17.73|4.35|

As we can see, the performance is mostly the same as my previous PR, until the hash table is many times larger than the amount of available memory. This is where this PR improves performance a lot. In the previous PR, when we had to do many partitioned probe phases, we created a lot of I/O pressure by reading and writing the entire probe side every time. In this PR, the I/O pressure is much less, as we are only reading and writing the probe side data _once_.

I will continue tweaking the performance in future PRs. Happy to receive feedback!

Edit: this PR uncovered a bug with how execution pipelines were scheduled. The problem was rather complicated, so I will try to explain it here.

We have parallelism in regular pipelines, e.g., `SCAN -> FILTER -> PROJECTION -> AGGREGATION`. The data is pushed from the source (`SCAN`) through the streaming `FILTER` and `PROJECTION` operators, into the sink (`AGGREGATION`).

Pipelines can also have dependencies on other pipelines, for example `SCAN -> JOIN` (build side of the join), and `SCAN -> JOIN -> AGGREGATION` (probe side of the join). The second pipeline depends on the first pipeline to be done.

Besides these, we also have two "special" cases of pipelines, namely "union" pipelines, for `UNION` queries, and for "child" pipelines, for streaming operators that become a source operator after streaming is done (e.g., scanning the hash table of a join for unmatched tuples for a `FULL OUTER` join). These cases are special because they share their sink with other pipelines.

We have run into problems with these special cases before because setting their dependencies up correctly is different from other pipelines. With the way the code was written before, it was very difficult to get things right. I've refactored pipeline construction with a new class called a `MetaPipeline`. This class holds multiple `Pipeline`s that share the same sink. This makes it much easier to set up dependencies correctly.

As a result (bonus!) of this refactor, we can now execute union pipelines in parallel, e.g.:
```sql
SELECT * FROM t1
UNION ALL
SELECT * FROM t2
UNION ALL
SELECT * FROM t3
```

Before we would scan t1, then t2, then t3. If t1, t2, and t3 are small enough to not keep all threads busy, we would not utilize all threads for the entire query. Now, we can scan all three tables concurrently.

Quick test with sorting a union:
```sql
create table test as select cast(random() * 100000000 as int) i from range(100000);
with union_cte as (
select i from test
union all
select i from test
union all
select i from test
union all
select i from test)
select count(i) from (select i from union_cte order by i offset 1);
```

This PR: 0.014s
Master: 0.023s

Commit:
init streaming partitioner

Commit:
initial implementation of PartitionedColumnData

Commit:
lock for shared column data allocator

Commit:
probe-side partitioning seems to (kind of) work!

Commit:
probe-side partitioning unit tests passing!

Commit:
chunk partitioning performance improvements

Commit:
offset pointer to sel vector instead of passing offset

Commit:
sort segments, ensure block size is BLOCK_SIZE

Commit:
Merge branch 'master' into oochj

Commit:
less locking and add ColumnDataAppendState to PartitionedColumnDataAppendState

Commit:
BLOCK_SIZE instead of BLOCK_ALLOC_SIZE, less tasks for ExternalBuild in external hash join

Commit:
add a bunch of TODO's so I don't forget

Commit:
Merge branch 'master' into oochj

Commit:
rework PartitionedColumnData API a bit

Commit:
some code cleanup

Commit:
Merge branch 'master' into oochj

Commit:
progress on scanning/consuming a ColumnDataCollection

Commit:
start work on ColumnDataConsumer

Commit:
properly implement ColumnDataConsumer for external hash join

Commit:
Merge branch 'master' into oochj

Commit:
always init probespill and re-enable anti/outer external joins

Commit:
delete block handles/pins more cleanly

Commit:
Merge branch 'master' into oochj

Commit:
enable out-of-core for anti/semi/mark joins

Commit:
add ProbeSpillLocalState to prevent some segfaultage

Commit:
simplify external GetData init

Commit:
Merge branch 'master' into oochj

Commit:
some code cleanup and trying to make CI happy

Commit:
Merge branch 'master' into oochj

Commit:
fix bug with CountValid (took me FOREVER to find out this was the problem)

Commit:
always init probespill

Commit:
trying to please windows CI

Commit:
fix race condition and nullptr

Commit:
Merge branch 'master' into oochj

Commit:
revert some change to prep for PR

Commit:
make BufferSize never exceed STANDARD_VECTOR_SIZE

Commit:
Merge branch 'master' into oochj

lnkuiper commented:
There seems to be one failing test on Windows 64 bit due to incorrect results: `test/sql/join/full_outer/full_outer_join_union.test`.

I've investigated this by manually running this test on Windows in Debug mode, and the assertion at `physical_hash_join.cpp:460` fails, meaning that `GetData` is called before the last call to `Execute` has happened.

This should have been fixed by #4340 and #4413, but somehow it is not, and the bug only shows up on Windows (I've tried running the test in a `while true; do ... ; done` loop for 10 minutes on my macbook, and it never fails.

Any ideas @Mytherin ?

Commit:
Merge branch 'master' into oochj

lnkuiper commented:
Just checked, and this also fails on master. Going to try and see if I fix the union pipeline scheduling

lnkuiper commented:
Just going to document my findings here, feel free to ignore.

The join where the assertion is triggered, is the top-level join in this physical plan:
```
┌───────────────────────────┐
│         HASH_JOIN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│            FULL           │
│           i = i           ├─────────────────────────────────────────────────────────────────────────────────────────────────────┐
│           j = j           │                                                                                                     │
│       EC = 0.000000       │                                                                                                     │
│      COST = 0.000000      │                                                                                                     │
└─────────────┬─────────────┘                                                                                                     │
┌─────────────┴─────────────┐                                                                                       ┌─────────────┴─────────────┐
│           UNION           ├───────────────────────────────────────────┐                                           │           UNION           ├───────────────────────────────────────────┐
└─────────────┬─────────────┘                                           │                                           └─────────────┬─────────────┘                                           │
┌─────────────┴─────────────┐                             ┌─────────────┴─────────────┐                             ┌─────────────┴─────────────┐                             ┌─────────────┴─────────────┐
│         HASH_JOIN         │                             │         HASH_JOIN         │                             │         HASH_JOIN         │                             │         HASH_JOIN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│            FULL           │                             │            FULL           │                             │            FULL           │                             │            FULL           │
│           i = k           ├──────────────┐              │           i = k           ├──────────────┐              │           i = k           ├──────────────┐              │           i = k           ├──────────────┐
│       EC = 0.000000       │              │              │       EC = 0.000000       │              │              │       EC = 0.000000       │              │              │       EC = 0.000000       │              │
│      COST = 0.000000      │              │              │      COST = 0.000000      │              │              │      COST = 0.000000      │              │              │      COST = 0.000000      │              │
└─────────────┬─────────────┘              │              └─────────────┬─────────────┘              │              └─────────────┬─────────────┘              │              └─────────────┬─────────────┘              │
┌─────────────┴─────────────┐┌─────────────┴─────────────┐┌─────────────┴─────────────┐┌─────────────┴─────────────┐┌─────────────┴─────────────┐┌─────────────┴─────────────┐┌─────────────┴─────────────┐┌─────────────┴─────────────┐
│          SEQ_SCAN         ││          SEQ_SCAN         ││          SEQ_SCAN         ││          SEQ_SCAN         ││          SEQ_SCAN         ││          SEQ_SCAN         ││          SEQ_SCAN         ││          SEQ_SCAN         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│          integers         ││         integers2         ││          integers         ││         integers2         ││          integers         ││         integers2         ││          integers         ││         integers2         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             i             ││             k             ││             i             ││             k             ││             i             ││             k             ││             i             ││             k             │
│             j             ││             l             ││             j             ││             l             ││             j             ││             l             ││             j             ││             l             │
│            EC=2           ││            EC=2           ││            EC=2           ││            EC=2           ││            EC=2           ││            EC=2           ││            EC=2           ││            EC=2           │
└───────────────────────────┘└───────────────────────────┘└───────────────────────────┘└───────────────────────────┘└───────────────────────────┘└───────────────────────────┘└───────────────────────────┘└───────────────────────────┘
```

The suspicion is that this has to do with union pipeline scheduling, so I tried to materialize all of the child operators of the unions, and run the query again. The assertion does not trigger if I do this.

So, it must have something to do with the combination of the unions, as well as the joins within the unions. I suspect what happens is that the `GetData` pipeline of the top-level join is dependent on the union pipelines that stream data through the `FULL HASH_JOIN` and through the union, but not dependent on the final pipeline (where that join is a source operator) that is scheduled after streaming is done.

Commit:
Merge branch 'master' into oochj

Commit:
trying to debug union pipeline scheduling

Commit:
add verification code for executor

Commit:
Merge branch 'master' into oochj

lnkuiper commented:
I've added a dependency between the top-level child pipeline and the child pipeline of the union pipeline, and seem to be running into a deadlock. So, I've added verification for circular event dependencies, and it does not seem to be due to a circular dependency.

However, now that I've added the verification code, there is no more deadlock, but it seems that the `PhysicalHashJoin` in the pipeline is being `free`'d before the pipeline is done, causing a different error. This cannot be due to the verification code as it does not manipulate the `Event`s (and I've `const`'d everything to be sure). I suspect that due to the verification code the timings are slightly different, and that therefore the behavior is different.

I don't understand why this is happening, just reporting my findings here.

Commit:
clean up verification code a bit (still struggling to find pipeline scheduling problem)

lnkuiper commented:
Seems I got confused while debugging this; it is a deadlock after all. My cyclic dependency checker for `Event`s was throwing an error, but due to some parallel stuff happening, the stack trace was from the pipeline being destroyed while another thread was trying to execute it.

So, good news, the dependency checker works, and I've found the cycle! Now I have to find out where it is being set ...

Commit:
initial implementation of MetaPipeline - not working yet

Commit:
add PipelineInitializeEvent and refactor so union pipelines can run in parallel

Commit:
refactor pipeline construction: got queries working again (somewhat)

Commit:
combinations of union/child pipelines seem to work - still some bugs to fix though

Commit:
fix recursive CTE after pipeline build refactor

Commit:
fix includes for faster compilation time, and fix union pipeline dependencies

Commit:
Merge branch 'master' into oochj

Commit:
pipeline build rework: just one test failing (recursive cte's grr)

Commit:
Merge branch 'master' into oochj

Commit:
only schedule recursive pipelines within PhysicalRecursiveCTE

Commit:
fix last issues with iejoin

Commit:
Merge branch 'master' into oochj

Commit:
fix bug with union all order preservation

Commit:
refactor IEJoin pipelines to have a single MetaPipeline

Commit:
properly set recursive CTE in MetaPipeline and add some missing includes

Commit:
trying to please the CI

Commit:
trying to please CI and dodge threadsan data races

Commit:
still trying to dodge threadsan data races

Commit:
Add lock to RandomEngine to make RandomInitLocalState thread-safe

Commit:
Merge branch 'master' into oochj

Commit:
add missing include

lnkuiper commented:
I think I got the final thread sanitizer issues out of the way, but the R + arrow test seems to be failing. Is this the test that occasionally fails, or is this something I caused? 10 tests fail with an std::exception

Mytherin commented:
Hm, these do not look like "the usual" spurious R test failures. At least I haven't seen them before. I will rerun the test.

lnkuiper commented:
Thanks for re-running the tests. This time only 4 failures, seems like the same problem though. Will investigate.

Mytherin commented on src/common/types/column_data_allocator.cpp:
```diff
@@ -29,7 +30,16 @@ ColumnDataAllocator::ColumnDataAllocator(ClientContext &context, ColumnDataAlloc
+	if (shared) {
+		lock_guard<mutex> guard(lock);
+		auto &block_handle = blocks[block_id].handle;
+		auto result = alloc.buffer_manager->Pin(block_handle);
```
nit: can we move this to a separate function (`PinInternal`)?

Mytherin commented on src/include/duckdb/execution/operator/order/physical_order.hpp:
```diff
@@ -33,6 +33,10 @@ class PhysicalOrder : public PhysicalOperator {
 	             LocalSourceState &lstate) const override;
 
+	bool IsOrderPreserving() const override {
+		return false;
```
Is this correct? Won't this lead to incorrect parallel result set materialization?

lnkuiper replied:
This is correct. It sounds confusing, but the `PhysicalOrder` does not preserve the insertion order - the data is completely reordered.

Mytherin commented on test/optimizer/statistics/statistics_setop.test:
```diff
@@ -11,6 +11,8 @@ CREATE TABLE integers2 AS SELECT * FROM (VALUES (4), (5), (6)) tbl(i);
 statement ok
 CREATE TABLE integers3 AS SELECT * FROM (VALUES (4), (5), (NULL)) tbl(i);
 
+mode skip
```
Can we re-enable all skipped tests?

Mytherin reviewed (commented):
Thanks for all the fixes! Looks great. The meta pipeline refactor looks very good. Some remaining comments from my side.

Commit:
reset source in main thread for R

Commit:
Merge branch 'master' into oochj

Commit:
implement PR feedback and fix skipped tests

lnkuiper commented:
Thanks for the feedback! I've implemented it. Together with Hannes, I've figured out what the R + arrow issue was. R does not like it if threads other than the main thread call back into R, which was what happened for arrow scans when I refactored the pipeline construction.

I've tried now to make sure that table function global source state initialization always happens in the main thread. Fingers crossed for CI 🤞

Commit:
Merge branch 'master' into oochj

Commit:
init global source state before calling continue in loop ... oops

Commit:
merge with master

Commit:
add missing return statement

lnkuiper commented:
Looks like my changes fixed the last failing R + arrow tests. I think the failures are unrelated. Let's wait for the full CI run though, which is taking forever because everyone is so productive!
